algebra/src/curves/mod.rs

```diff
@@ -56,274 +23,106 @@ pub mod tests;
-        G2: Into<Self::G2Affine>,
-    {
-        let g1_prep = Self::G1Prepared::from(p.into());
-        let g2_prep = Self::G2Prepared::from(q.into());
-        Self::product_of_pairings(std::iter::once(&(g1_prep, g2_prep)))
+    #[inline]
+    fn from_affine<'a>(other: &'a Self::AffineRep) -> Self {
+        Into::<Self>::into(*other)
     }
-}
 
-/// Projective representation of an elliptic curve point guaranteed to be
-/// in the correct prime order subgroup.
-pub trait ProjectiveCurve:
-    Eq
-    + Sized
-    + ToBytes
-    + FromBytes
-    + Serialize
-    + for<'a> Deserialize<'a>
-    + CanonicalSerialize
-    + CanonicalDeserialize
-    + SemanticallyValid
-    + FromBytesChecked
-    + Copy
-    + Clone
-    + Default
-    + Send
-    + Sync
-    + Hash
-    + Debug
-    + Display
-    + UniformRand
-    + 'static
-    + Neg<Output = Self>
-    + for<'a> Add<&'a Self, Output = Self>
-    + for<'a> Sub<&'a Self, Output = Self>
-    + for<'a> AddAssign<&'a Self>
-    + for<'a> SubAssign<&'a Self>
-{
-    type ScalarField: PrimeField + SquareRootField + Into<<Self::ScalarField as PrimeField>::BigInt>;
-    type BaseField: Field;
-    type Affine: AffineCurve<Projective = Self, ScalarField = Self::ScalarField>;
-
-    /// Returns the additive identity.
-    #[must_use]
-    fn zero() -> Self;
-
-    /// Returns a fixed generator of unknown exponent.
-    #[must_use]
-    fn prime_subgroup_generator() -> Self;
-
-    /// Determines if this point is the point at infinity.
-    #[must_use]
-    fn is_zero(&self) -> bool;
-
-    /// Checks that the current point is on curve and is in the
-    /// prime order subgroup
-    #[must_use]
-    fn group_membership_test(&self) -> bool;
-
-    /// Normalizes a slice of projective elements so that
-    /// conversion to affine is cheap.
-    fn batch_normalization(v: &mut [Self]);
+    fn batch_from_affine<'a>(vec_affine: &'a [Self::AffineRep]) -> Vec<Self> {
+        vec_affine
+            .iter()
+            .map(|&affine| affine.into())
+            .collect::<Vec<_>>()
+    }
 
-    fn batch_normalization_into_affine(mut v: Vec<Self>) -> Vec<Self::Affine> {
-        Self::batch_normalization(v.as_mut_slice());
-        v.into_iter().map(|p| p.into_affine()).collect()
+    fn batch_into_affine<'a>(vec_self: &'a [Self]) -> Result<Vec<Self::AffineRep>, Error> {
+        vec_self
+            .iter()
+            .map(|&projective| projective.into_affine())
+            .collect::<Result<Vec<_>, _>>()
     }
 
-    /// Checks if the point is already "normalized" so that
-    /// cheap affine conversion is possible.
-    #[must_use]
-    fn is_normalized(&self) -> bool;
+    fn add_affine<'a>(&self, other: &'a Self::AffineRep) -> Self;
 
-    /// Doubles this element.
-    #[must_use]
-    fn double(&self) -> Self {
-        let mut copy = *self;
-        copy.double_in_place();
-        copy
-    }
+    fn add_affine_assign<'a>(&mut self, other: &'a Self::AffineRep);
```

Comment:
Can't we use Rust std Add traits to be mixed into the Curve trait ?

Reply:
This is more question to @UlrichHaboeck75, his opinion is that better to have a separate function with specific name for mixed operations. Of course we can implement Add / AddAssign traits for pair "G + G::AffineRep"

Reply:
Honestly, to me, it's more elegant (and best practice) to use Rust traits

Reply:
I agree, from a high-level perspective it would be more comfortable to implement the `Add` and `AddAssign` trait. However, I would not hide the usage of mixed type arithmetics behind a trait with generic names. Mixed type arithemtics are very specific and typically used to improve performance over the generic group implementation. For code review (and performance improvement tasks) I believe it is important to point out in code where it is used and where not.

Reply:
Then let's rename the functions like they were before (add_affine_mixed)

Reply:
It was a quite long process of clarifying the Curve trait, and choosing meaningful names for its functions.
Is there a specific reason for renaming? If it is just about taste, I wouldn't recommend to change.

Reply:
I've left them as they were (just renamed add_affine_assign into add_assign_affine, as per standard nomenclature)